App ServiceでマネージドIDのCredentialsを読み込めない

# 概要
新規登録画面から画像を投稿できない。
投稿する画像ファイルの書き込み先にBLOBストレージを設定しており、DefaultAzureCredentialによって認証している。
App Serviceにはユーザー割り当てマネージドIDにより以下のの権限を割り当てている。
- Storage Blob Data Contributor
- AcrPull
- Key Vault Secrets User

ACRからのイメージPULLやKeyVaultのシークレット参照はできる。
回避策として、App Serviceの環境変数にサービスプリンシパルのCredentialsを追加すると正常に動作する。
```
AZURE_CLIENT_ID
AZURE_TENANT_ID
AZURE_CLIENT_SECRET
```

認証の実装部分は以下。`AZURE_TOKEN_CREDENTIAL = DefaultAzureCredential()`で認証している。
```production.py
import sys
from .base import *
from .base import env
from azure.identity import DefaultAzureCredential
from opencensus.ext.azure.log_exporter import AzureLogHandler
:
# STORAGES
INSTALLED_APPS += ["storages"]
AZURE_ACCOUNT_NAME = env("DJANGO_AZURE_ACCOUNT_NAME")
AZURE_STATIC_CONTAINER = env("DJANGO_AZURE_STATIC_CONTAINER")
AZURE_MEDIA_CONTAINER = env("DJANGO_AZURE_MEDIA_CONTAINER")
AZURE_TOKEN_CREDENTIAL = DefaultAzureCredential()
AZURE_CUSTOM_DOMAIN = env("DJANGO_ALLOWED_HOSTS")
:
```

[DefaultAzureCredential](https://learn.microsoft.com/ja-jp/azure/developer/python/sdk/authentication-overview#use-defaultazurecredential-in-an-application) は以下の優先順位で認証するはず。
- 環境変数
- マネージドID
- Azure CLI
- Azure PowerShell
- 対話形式で認証

# 環境
- Python==3.9.2
- Django==4.0.2
- azure-identity==1.12.0
- django-storages==1.13.2


# エラーメッセージ
```
2023-05-10T07:14:13.146698690Z: [INFO]  Attempted credentials:

2023-05-10T07:14:13.146702590Z: [INFO]  	EnvironmentCredential: EnvironmentCredential authentication unavailable. Environment variables are not fully configured.

2023-05-10T07:14:13.146717490Z: [INFO]  Visit https://aka.ms/azsdk/python/identity/environmentcredential/troubleshoot to troubleshoot.this issue.

2023-05-10T07:14:13.146722490Z: [INFO]  	ManagedIdentityCredential: (None) Unable to load the proper Managed Identity.

2023-05-10T07:14:13.146726691Z: [INFO]  Code: None

2023-05-10T07:14:13.146730691Z: [INFO]  Message: Unable to load the proper Managed Identity.

2023-05-10T07:14:13.146734791Z: [INFO]  To mitigate this issue, please refer to the troubleshooting guidelines here at https://aka.ms/azsdk/python/identity/defaultazurecredential/troubleshoot.

2023-05-10T07:14:13.344226846Z: [INFO]  2023-05-10 16:14:13,147 [ERROR] 20 132483186174776 /usr/local/lib/python3.9/site-packages/django/utils/log.py:225 Internal Server Error: /create/

2023-05-10T07:14:13.344260747Z: [INFO]  Traceback (most recent call last):

2023-05-10T07:14:13.344266447Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 47, in inner

2023-05-10T07:14:13.344270747Z: [INFO]      response = get_response(request)

2023-05-10T07:14:13.344274347Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 181, in _get_response

2023-05-10T07:14:13.344285047Z: [INFO]      response = wrapped_callback(request, *callback_args, **callback_kwargs)

2023-05-10T07:14:13.344290047Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/views/generic/base.py", line 69, in view

2023-05-10T07:14:13.344293747Z: [INFO]      return self.dispatch(request, *args, **kwargs)

2023-05-10T07:14:13.344297647Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/contrib/auth/mixins.py", line 71, in dispatch

2023-05-10T07:14:13.344301447Z: [INFO]      return super().dispatch(request, *args, **kwargs)

2023-05-10T07:14:13.344305047Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/views/generic/base.py", line 101, in dispatch

2023-05-10T07:14:13.344308747Z: [INFO]      return handler(request, *args, **kwargs)

2023-05-10T07:14:13.344312248Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/views/generic/edit.py", line 174, in post

2023-05-10T07:14:13.344315848Z: [INFO]      return super().post(request, *args, **kwargs)

2023-05-10T07:14:13.344319348Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/views/generic/edit.py", line 144, in post

2023-05-10T07:14:13.344322948Z: [INFO]      return self.form_valid(form)

2023-05-10T07:14:13.344326448Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/views/generic/edit.py", line 127, in form_valid

2023-05-10T07:14:13.344330048Z: [INFO]      self.object = form.save()

2023-05-10T07:14:13.344333548Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/forms/models.py", line 466, in save

2023-05-10T07:14:13.344337148Z: [INFO]      self.instance.save()

2023-05-10T07:14:13.344340548Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 743, in save

2023-05-10T07:14:13.344344148Z: [INFO]      self.save_base(using=using, force_insert=force_insert,

2023-05-10T07:14:13.344357648Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 780, in save_base

2023-05-10T07:14:13.344361948Z: [INFO]      updated = self._save_table(

2023-05-10T07:14:13.344365348Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 885, in _save_table

2023-05-10T07:14:13.344368848Z: [INFO]      results = self._do_insert(cls._base_manager, using, fields, returning_fields, raw)

2023-05-10T07:14:13.344372448Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 923, in _do_insert

2023-05-10T07:14:13.344377348Z: [INFO]      return manager._insert(

2023-05-10T07:14:13.344380948Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method

2023-05-10T07:14:13.344384548Z: [INFO]      return getattr(self.get_queryset(), name)(*args, **kwargs)

2023-05-10T07:14:13.344387948Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 1301, in _insert

2023-05-10T07:14:13.344391548Z: [INFO]      return query.get_compiler(using=using).execute_sql(returning_fields)

2023-05-10T07:14:13.344395048Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1440, in execute_sql

2023-05-10T07:14:13.344399049Z: [INFO]      for sql, params in self.as_sql():

2023-05-10T07:14:13.344402549Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1382, in as_sql

2023-05-10T07:14:13.344406149Z: [INFO]      value_rows = [

2023-05-10T07:14:13.344409549Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1383, in <listcomp>

2023-05-10T07:14:13.344413649Z: [INFO]      [self.prepare_value(field, self.pre_save_val(field, obj)) for field in fields]

2023-05-10T07:14:13.344417149Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1383, in <listcomp>

2023-05-10T07:14:13.344420849Z: [INFO]      [self.prepare_value(field, self.pre_save_val(field, obj)) for field in fields]

2023-05-10T07:14:13.344424349Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1334, in pre_save_val

2023-05-10T07:14:13.344427949Z: [INFO]      return field.pre_save(obj, add=True)

2023-05-10T07:14:13.344431349Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/db/models/fields/files.py", line 302, in pre_save

2023-05-10T07:14:13.344434949Z: [INFO]      file.save(file.name, file.file, save=False)

2023-05-10T07:14:13.344438449Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/db/models/fields/files.py", line 89, in save

2023-05-10T07:14:13.344441949Z: [INFO]      self.name = self.storage.save(name, content, max_length=self.field.max_length)

2023-05-10T07:14:13.344457349Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/core/files/storage.py", line 53, in save

2023-05-10T07:14:13.344461149Z: [INFO]      name = self.get_available_name(name, max_length=max_length)

2023-05-10T07:14:13.344464849Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/storages/backends/azure_storage.py", line 259, in get_available_name

2023-05-10T07:14:13.344472349Z: [INFO]      return super().get_available_name(name, max_length)

2023-05-10T07:14:13.344476749Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/django/core/files/storage.py", line 91, in get_available_name

2023-05-10T07:14:13.344480550Z: [INFO]      while self.exists(name) or (max_length and len(name) > max_length):

2023-05-10T07:14:13.344484850Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/storages/backends/azure_storage.py", line 263, in exists

2023-05-10T07:14:13.344488650Z: [INFO]      return blob_client.exists()

2023-05-10T07:14:13.344492350Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/core/tracing/decorator.py", line 76, in wrapper_use_tracer

2023-05-10T07:14:13.344508150Z: [INFO]      return func(*args, **kwargs)

2023-05-10T07:14:13.344511950Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/storage/blob/_blob_client.py", line 1241, in exists

2023-05-10T07:14:13.344515850Z: [INFO]      process_storage_error(error)

2023-05-10T07:14:13.344519650Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/storage/blob/_shared/response_handlers.py", line 97, in process_storage_error

2023-05-10T07:14:13.344523650Z: [INFO]      raise storage_error

2023-05-10T07:14:13.344527450Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/storage/blob/_blob_client.py", line 1232, in exists

2023-05-10T07:14:13.344531350Z: [INFO]      self._client.blob.get_properties(

2023-05-10T07:14:13.344535150Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/core/tracing/decorator.py", line 76, in wrapper_use_tracer

2023-05-10T07:14:13.344539050Z: [INFO]      return func(*args, **kwargs)

2023-05-10T07:14:13.344542850Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/storage/blob/_generated/operations/_blob_operations.py", line 1874, in get_properties

2023-05-10T07:14:13.344546850Z: [INFO]      pipeline_response = self._client._pipeline.run(  # type: ignore # pylint: disable=protected-access

2023-05-10T07:14:13.344550750Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 202, in run

2023-05-10T07:14:13.344554650Z: [INFO]      return first_node.send(pipeline_request)

2023-05-10T07:14:13.344558450Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 70, in send

2023-05-10T07:14:13.344562350Z: [INFO]      response = self.next.send(request)

2023-05-10T07:14:13.344566151Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 70, in send

2023-05-10T07:14:13.344570051Z: [INFO]      response = self.next.send(request)

2023-05-10T07:14:13.344573851Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 70, in send

2023-05-10T07:14:13.344577751Z: [INFO]      response = self.next.send(request)

2023-05-10T07:14:13.344581451Z: [INFO]    [Previous line repeated 2 more times]

2023-05-10T07:14:13.344585251Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/policies/_redirect.py", line 156, in send

2023-05-10T07:14:13.344589251Z: [INFO]      response = self.next.send(request)

2023-05-10T07:14:13.344592951Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 70, in send

2023-05-10T07:14:13.344600251Z: [INFO]      response = self.next.send(request)

2023-05-10T07:14:13.344604251Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/storage/blob/_shared/policies.py", line 546, in send

2023-05-10T07:14:13.344608251Z: [INFO]      raise err

2023-05-10T07:14:13.344612351Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/storage/blob/_shared/policies.py", line 520, in send

2023-05-10T07:14:13.344616351Z: [INFO]      response = self.next.send(request)

2023-05-10T07:14:13.344620151Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 70, in send

2023-05-10T07:14:13.344624051Z: [INFO]      response = self.next.send(request)

2023-05-10T07:14:13.344627851Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 70, in send

2023-05-10T07:14:13.344631751Z: [INFO]      response = self.next.send(request)

2023-05-10T07:14:13.344635451Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/policies/_authentication.py", line 108, in send

2023-05-10T07:14:13.344639351Z: [INFO]      self.on_request(request)

2023-05-10T07:14:13.344643151Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/policies/_authentication.py", line 87, in on_request

2023-05-10T07:14:13.344647052Z: [INFO]      self._token = self._credential.get_token(*self._scopes)

2023-05-10T07:14:13.344650852Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/identity/_credentials/default.py", line 168, in get_token

2023-05-10T07:14:13.344654852Z: [INFO]      return super(DefaultAzureCredential, self).get_token(*scopes, **kwargs)

2023-05-10T07:14:13.344658652Z: [INFO]    File "/usr/local/lib/python3.9/site-packages/azure/identity/_credentials/chained.py", line 101, in get_token

2023-05-10T07:14:13.344662652Z: [INFO]      raise ClientAuthenticationError(message=message)

2023-05-10T07:14:13.344666352Z: [INFO]  azure.core.exceptions.ClientAuthenticationError: DefaultAzureCredential failed to retrieve a token from the included credentials.

2023-05-10T07:14:13.344670352Z: [INFO]  Attempted credentials:

2023-05-10T07:14:13.344674052Z: [INFO]  	EnvironmentCredential: EnvironmentCredential authentication unavailable. Environment variables are not fully configured.

2023-05-10T07:14:13.344678052Z: [INFO]  Visit https://aka.ms/azsdk/python/identity/environmentcredential/troubleshoot to troubleshoot.this issue.

2023-05-10T07:14:13.344681952Z: [INFO]  	ManagedIdentityCredential: (None) Unable to load the proper Managed Identity.

2023-05-10T07:14:13.344685852Z: [INFO]  Code: None

2023-05-10T07:14:13.344689552Z: [INFO]  Message: Unable to load the proper Managed Identity.
```

--- a/terraform/keyvault.tf
+++ b/terraform/keyvault.tf
@@ -43,24 +43,6 @@ resource "azurerm_key_vault_secret" "default_from_email" {
   key_vault_id = azurerm_key_vault.app.id
 }
 
-resource "azurerm_key_vault_secret" "azure_client_id" {
-  name         = "AZURE-CLIENT-ID"
-  value        = var.azure_client_id
-  key_vault_id = azurerm_key_vault.app.id
-}
-
-resource "azurerm_key_vault_secret" "azure_tenant_id" {
-  name         = "AZURE-TENANT-ID"
-  value        = var.azure_tenant_id
-  key_vault_id = azurerm_key_vault.app.id
-}
-
-resource "azurerm_key_vault_secret" "azure_client_secret" {
-  name         = "AZURE-CLIENT-SECRET"
-  value        = var.azure_client_secret
-  key_vault_id = azurerm_key_vault.app.id
-}
-
 # Azure Cache for Redis
 resource "azurerm_key_vault_secret" "redis_host" {
   name         = "REDIS-HOST"

--- a/terraform/keyvault.tf
+++ b/terraform/keyvault.tf
@@ -43,6 +43,12 @@ resource "azurerm_key_vault_secret" "default_from_email" {
   key_vault_id = azurerm_key_vault.app.id
 }
 
+resource "azurerm_key_vault_secret" "webappcontainer_client_id" {
+  name         = "WEBAPPCONTAINER-CLIENT-ID"
+  value        = azurerm_user_assigned_identity.webappcontainer.client_id
+  key_vault_id = azurerm_key_vault.app.id
+}
+
 # Azure Cache for Redis
 resource "azurerm_key_vault_secret" "redis_host" {
   name         = "REDIS-HOST"

--- a/terraform/variables.tf
+++ b/terraform/variables.tf
@@ -32,16 +32,6 @@ variable "default_from_email" {
   type = string
 }
 
-variable "azure_client_id" {
-  type = string
-}
-variable "azure_tenant_id" {
-  type = string
-}
-variable "azure_client_secret" {
-  type = string
-}
-
 # Azure Storage
 variable "account_replication_type" {
   type    = string

--- a/terraform/webappcontainer.tf
+++ b/terraform/webappcontainer.tf
@@ -55,9 +55,6 @@ resource "azurerm_linux_web_app" "app" {
     "DJANGO_DEFAULT_FROM_EMAIL"           = "@Microsoft.KeyVault(VaultName=${azurerm_key_vault.app.name};SecretName=DEFAULT-FROM-EMAIL)"
     "DJANGO_SECRET_KEY"                   = "@Microsoft.KeyVault(VaultName=${azurerm_key_vault.app.name};SecretName=DJANGO-SECRET-KEY)"
     "APPINSIGHTS_CONNECTION_STRING"       = "@Microsoft.KeyVault(VaultName=${azurerm_key_vault.app.name};SecretName=APPINSIGHTS-CONNECTION-STRING)"
-    "AZURE_CLIENT_ID"                     = "@Microsoft.KeyVault(VaultName=${azurerm_key_vault.app.name};SecretName=AZURE-CLIENT-ID)"
-    "AZURE_TENANT_ID"                     = "@Microsoft.KeyVault(VaultName=${azurerm_key_vault.app.name};SecretName=AZURE-TENANT-ID)"
-    "AZURE_CLIENT_SECRET"                 = "@Microsoft.KeyVault(VaultName=${azurerm_key_vault.app.name};SecretName=AZURE-CLIENT-SECRET)"
   }
 
   # Use Key Vault references for App Service

--- a/terraform/webappcontainer.tf
+++ b/terraform/webappcontainer.tf
@@ -55,6 +55,7 @@ resource "azurerm_linux_web_app" "app" {
     "DJANGO_DEFAULT_FROM_EMAIL"           = "@Microsoft.KeyVault(VaultName=${azurerm_key_vault.app.name};SecretName=DEFAULT-FROM-EMAIL)"
     "DJANGO_SECRET_KEY"                   = "@Microsoft.KeyVault(VaultName=${azurerm_key_vault.app.name};SecretName=DJANGO-SECRET-KEY)"
     "APPINSIGHTS_CONNECTION_STRING"       = "@Microsoft.KeyVault(VaultName=${azurerm_key_vault.app.name};SecretName=APPINSIGHTS-CONNECTION-STRING)"
+    "AZURE_CLIENT_ID"                     = "@Microsoft.KeyVault(VaultName=${azurerm_key_vault.app.name};SecretName=WEBAPPCONTAINER-CLIENT-ID)"
   }
 
   # Use Key Vault references for App Service

